ako-gateway-api/k8s/validator.go

```diff
@@ -188,32 +244,51 @@ func isValidListener(key string, gateway *gatewayv1.Gateway, gatewayStatus *gate
 			Reason(string(gatewayv1.ListenerReasonUnsupportedProtocol)).
 			Message("Unsupported protocol").
 			SetIn(&gatewayStatus.Listeners[index].Conditions)
+		programmedCondition.SetIn(&gatewayStatus.Listeners[index].Conditions)
 		gatewayStatus.Listeners[index].SupportedKinds = akogatewayapilib.SupportedKinds[gatewayv1.HTTPSProtocolType]
 		return false
 	}
 
+	resolvedRefCondition := akogatewayapistatus.NewCondition().
+		Type(string(gatewayv1.ListenerConditionResolvedRefs)).
+		Status(metav1.ConditionFalse).
+		ObservedGeneration(gateway.ObjectMeta.Generation)
 	// has valid TLS config
 	if listener.TLS != nil {
 		if (listener.TLS.Mode != nil && *listener.TLS.Mode != gatewayv1.TLSModeTerminate) || len(listener.TLS.CertificateRefs) == 0 {
 			utils.AviLog.Errorf("key: %s, msg: tls mode/ref not valid %+v/%+v", key, gateway.Name, listener.Name)
-			defaultCondition.
-				Reason(string(gatewayv1.ListenerReasonInvalidCertificateRef)).
+			defaultCondition.SetIn(&gatewayStatus.Listeners[index].Conditions)
+			resolvedRefCondition.Reason(string(gatewayv1.ListenerReasonInvalidCertificateRef)).
 				Message("TLS mode or reference not valid").
 				SetIn(&gatewayStatus.Listeners[index].Conditions)
+			programmedCondition.SetIn(&gatewayStatus.Listeners[index].Conditions)
 			return false
 		}
 		for _, certRef := range listener.TLS.CertificateRefs {
 			//only secret is allowed
 			if (certRef.Group != nil && string(*certRef.Group) != "") ||
 				certRef.Kind != nil && string(*certRef.Kind) != utils.Secret {
 				utils.AviLog.Errorf("key: %s, msg: CertificateRef is not valid %+v/%+v, must be Secret", key, gateway.Name, listener.Name)
-				defaultCondition.
-					Reason(string(gatewayv1.ListenerReasonInvalidCertificateRef)).
+				defaultCondition.SetIn(&gatewayStatus.Listeners[index].Conditions)
+				resolvedRefCondition.Reason(string(gatewayv1.ListenerReasonInvalidCertificateRef)).
 					Message("TLS mode or reference not valid").
 					SetIn(&gatewayStatus.Listeners[index].Conditions)
+				programmedCondition.SetIn(&gatewayStatus.Listeners[index].Conditions)
+				return false
+			}
+			name := string(certRef.Name)
+			cs := utils.GetInformers().ClientSet
+			secretObj, err := cs.CoreV1().Secrets(gateway.ObjectMeta.Namespace).Get(context.TODO(), name, metav1.GetOptions{})
+			if err != nil || secretObj == nil {
```

Comment:
we can take out `secretObj == nil` condition as if secret doesn't exist, it will be err always. Can be taken up in next PR.